### ISSUE TITLE
`prefer-for-of` false positive on interface with length/indexer

### ISSUE THREAD

**Issue by nchen63** [palantir/tslint#2021]
### Bug Report

- __TSLint version__: `4.3.1`
- __TypeScript version__: `2.1.4`
- __Running TSLint via__: (please choose one) vscode

#### TypeScript code being linted

```ts

interface FileList {
    readonly length: number;
    item(index: number): File;
    [index: number]: File;
}

// fileList is an instance of a class that implements FileList
for (let i = 0; i < fileList.length; i++) {
}
```

with `tslint.json` configuration:

```json
prefer-for-of: true
```

#### Actual behavior

indicates that for loop should be for-of

#### Expected behavior

`fileList` is not actually an array and TypeScript errors if you try to make it use for-of

**Comment by andy-hanson** [comment]
This should wait on Microsoft/TypeScript#13502. We would need to check assignability to `Iterable`.

**Comment by ajafff** [comment]
FWIW I have a working implementation of `prefer-for-of` with type information as part of the core rules of the [Wotan linter runtime](https://github.com/fimbullinter/wotan/blob/master/packages/wotan/README.md).

**Comment by JoshuaKGoldberg** [comment]
### 💀 _**It's time!**_ 💀

**TSLint is deprecated** and no longer accepting pull requests other than security fixes. See #4534. ☠️ 
We recommend you instead use **[typescript-eslint](https://typescript-eslint.io)** to lint your TypeScript code with **ESLint**. ✅

👋 It was a pleasure open sourcing with you!

**Comment by JoshuaKGoldberg** [comment]
🤖 Beep boop! 👉  **TSLint is deprecated** 👈 _(#4534)_  and you should **switch to [typescript-eslint](https://typescript-eslint.io)**! 🤖 

🔒 This issue is being locked to prevent further unnecessary discussions. Thank you! 👋
